src/analyses/base.ml

```diff
@@ -781,14 +832,29 @@ struct
     with IntDomain.ArithmeticOnIntegerBot _ ->
     ValueDomain.Compound.top_value (typeOf exp)
 
+  let query_evalint ask gs st e =
+    if M.tracing then M.traceli "evalint" "base query_evalint %a\n" d_exp e;
+    let r = match eval_rv_no_ask_evalint ask gs st e with
+    | `Int i -> i (* TODO: cast to right ikind here? or is it guaranteed? *)
+    | `Bot   -> Queries.ID.bot () (* TODO: remove? *)
+    (* | v      -> M.warn ("Query function answered " ^ (VD.show v)); Queries.Result.top q *)
+    | v      -> M.warn ("Query function answered " ^ (VD.show v)); Queries.ID.bot ()
+    in
+    if M.tracing then M.traceu "evalint" "base query_evalint %a -> %a\n" d_exp e Queries.ID.pretty r;
+    r
+
   (* Evaluate an expression containing only locals. This is needed for smart joining the partitioned arrays where ctx is not accessible. *)
   (* This will yield `Top for expressions containing any access to globals, and does not make use of the query system. *)
   (* Wherever possible, don't use this but the query system or normal eval_rv instead. *)
-  let eval_exp x (exp:exp) =
+  let eval_exp st (exp:exp) =
     (* Since ctx is not available here, we need to make some adjustments *)
-    let knownothing = { Queries.f = fun (type a) (q: a Queries.t) -> Queries.Result.top q } in (* our version of ask *)
-    let gs = fun _ -> G.top () in (* the expression is guaranteed to not contain globals *)
-    match (eval_rv knownothing gs x exp) with
+    let rec query: type a. a Queries.t -> a Queries.result = fun q ->
+      match q with
+      | EvalInt e -> query_evalint ask gs st e (* mimic EvalInt query since eval_rv needs it *)
```

Comment:
I'll leave this comment here on how it works for more complicated expressions, because it took me a while to follow the logic there.

```
E.g. EvalInt ((x+y)+z)
-> query_evalint ((x+y)+z)
   -> eval_rv_no_ask_evalint ((x+y)+z)
     -> eval_rv_ask_mustbeequal ((x+y)+z) (*)
       -> eval_rv_base ((x+y)+z)
         -> eval_rv (x+y) (**)
         -> eval_rv (z)
```

(*) and (**) use the ask that we provide that falls through to `eval_rv_base` all the time.